InternalDocs/garbage_collector.md

```diff
@@ -351,38 +351,82 @@ follows these steps in order:
+
+Generational garbage collection takes advantage of what is known as the weak
+generational hypothesis: Most objects die young.
+This has proven to be very close to the reality of many Python
 programs as many temporary objects are created and destroyed very quickly.
 
 To take advantage of this fact, all container objects are segregated into
-three spaces/generations. Every new
-object starts in the first generation (generation 0). The previous algorithm is
-executed only over the objects of a particular generation and if an object
-survives a collection of its generation it will be moved to the next one
-(generation 1), where it will be surveyed for collection less often. If
-the same object survives another GC round in this new generation (generation 1)
-it will be moved to the last generation (generation 2) where it will be
-surveyed the least often.
-
-The GC implementation for the free-threaded build does not use multiple
-generations.  Every collection operates on the entire heap.
+two generations: young and old. Every new object starts in the young generation.
+In order to keep pause times down, scanning of the old generation of the heap
+occurs in increments. To keep track of what has been scanned,
+the old generation contains two lists:
+
+* Those objects that have not yet been scanned, referred to as the `pending` list.
+* Those objects that have been scanned, referred to as the `visited` list.
+
+To detect and collect all unreachable objects in the heap, the garbage collector
+must scan the whole heap. This whole heap scan is called a full scavenge.
+
+Increments
+----------
+
+Each full scavenge is performed in a series of increments.
+For each full scavenge, the combined increments will cover the whole heap.
+
+For each increment, the portion of the heap scanned by a single collection is
+made up of three parts:
+
+* The young generation
+* The old generation's least recently scanned objects
+* All objects reachable from those objects that have not yet been scanned this full scavenge
+
+The surviving objects (those that are not collected) are moved to the back of the
+`visited` list in the old generation.
+
+When a full scavenge starts, no objects in the heap are considered to have been scanned,
+so all objects in the old generation must be in the `pending` space.
+When all objects in the heap have been scanned a cycle ends, and all objects are moved
+to the `pending` list again. To avoid having to traverse the entire list, which list is
+`pending` and which is `visited` is determined by a field in the `GCState` struct.
+The `visited` and `pending` lists can be swapped by toggling this bit.
+
+Correctness
+-----------
+
+In order to collect all unreachable cycles, each increment must contain all of
+an unreachable cycle, or none of it.
+In order to make sure that the whole of any unreachable cycle is contained in an
+increment,  all unscanned objects reachable from any object in the increment must
```

Comment:
I'm finding this a bit confusing - when we do an increment, we start at a random unscathed object (right?) and the find everything reachable from it which is not yet scanned, and then how do we know that this is an unreachable cycle?

Reply:
We don't. But if any object in the increment is part of such a cycle, then the whole cycle must be part of the increment.
Finding the cycles is the job of the base algorithm, we just need to be sure that we don't include part of a cycle in an increment.